### ISSUE TITLE
Create Custom Attributes in cognito from amplify cli

### ISSUE THREAD

**Issue by achutkiran** [aws-amplify/amplify-cli#1204]
>**Note**: If your question is regarding the AWS Amplify Console service, please log it in the 
[official AWS Amplify Console forum](https://forums.aws.amazon.com/forum.jspa?forumID=314&start=0)

** Which Category is your question related to? **
Amplify Auth

** What AWS Services are you utilizing? **
Cognito

** Provide additional details e.g. code snippets **
Is there a way to create custom attributes in cognito during 
`amplify auth add`
or by editing parameters.json

**Comment by kaustavghosh06** [comment]
@achutkiran Are you referring to the social provider custom attributes?

**Comment by achutkiran** [comment]
User pool custom attributes  like customer type

**Comment by jarrettj** [comment]
+1

**Comment by timkuilman** [comment]
You can manually change the `cognito-cloudformation-template.yml` file after adding the authentication (but before running `amplify push`.

[AWS properties Cognito UserPool](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-cognito-userpool-schemaattribute.html)


_Snippet cognito-cloudformation-template.yml_
```   
....
# BEGIN USER POOL RESOURCES
  UserPool:
  # Created upon user selection
  # Depends on SNS Role for Arn if MFA is enabled
    Type: AWS::Cognito::UserPool
    UpdateReplacePolicy: Retain
    Properties:
      UserPoolName: !If [ShouldNotCreateEnvResources, !Ref userPoolName, !Join ['',[!Ref userPoolName, '-', !Ref env]]]
      Schema:
        - Name: myCustomAttribute
          Required: true
          Mutable: true
          AttributeDataType: String

        - Name: myOtherCustomAttribute
          Required: false
          Mutable: true
          AttributeDataType: String

      AutoVerifiedAttributes: !Ref autoVerifiedAttributes
      Policies:
        PasswordPolicy:
          MinimumLength: !Ref passwordPolicyMinLength
          RequireLowercase: true
          RequireNumbers: true
          RequireSymbols: true
          RequireUppercase: true
      MfaConfiguration: !Ref mfaConfiguration
      SmsVerificationMessage: !Ref smsVerificationMessage
      SmsConfiguration:
        SnsCallerArn: !GetAtt SNSRole.Arn
        ExternalId: !Ref roleExternalId
...

```

**Comment by achutkiran** [comment]
Presently I am editing cloudFormationTemplate and parameters.json to add custom Attributes and groups. But problem is for any reason whenever i update auth i have to update template manually. So it will be better if custom Attributes are included in amplify cli.

**Comment by kaustavghosh06** [comment]
@achutkiran You're right, we don't have in-built support for custom attributes yet and what you're trying out is currently the best way to go about it, but I'll tag this as a feture request and prioritize work on this.

**Comment by jarrettj** [comment]
@timkuilman @achutkiran Yes I figured that out. I prefer editing the CF templates as well. Been a while since I've been in there. But not too hard to follow. And it would be impossible to have all the AWS service options in the cli anyways. Thanks.

**Comment by achutkiran** [comment]
Thanks

**Comment by yukitaka13-1110** [comment]
I tried https://github.com/aws-amplify/amplify-cli/issues/1204#issuecomment-482028480 and update auth, but an update error occured.

cloudformation.yml
```yml
        - Name: custom:store_name
          Required: true
          Mutable: true
          AttributeDataType: String
```

parameter.json
```json
    "requiredAttributes": [
        "...",
        "custom:store_name"
    ],
```

Am I wrong in specifying?

**Comment by RossWilliams** [comment]
@yukitaka13-1110 required attributes in Cognito must be defined in Cloudformation templates when you first create the user pool. Cognito does not allow this update. Here is a forum thread on the topic: https://forums.aws.amazon.com/thread.jspa?messageID=898280&#898280

**Comment by yukitaka13-1110** [comment]
@RossWilliams I tried https://github.com/aws-amplify/amplify-cli/issues/1204#issuecomment-482028480.
In this case, dose the ```myCustomAttribute``` stands for custom attribute(ex user_type) ?
about https://github.com/aws-amplify/amplify-cli/issues/1761#issuecomment-507007122  
I want to add custom attributes like user_type instead of required attribute.
Is it possible to add custom attributes by adding custom attributes in the cloudformation template and 'amplify push'?

**Comment by jarrettj** [comment]
@yukitaka13-1110 An example https://github.com/aws-amplify/amplify-cli/pull/1289/commits/83476b944852ffa000875074e200b4fd5ec6f88c?short_path=ac1bfc4#diff-ac1bfc471563fa82f8882dccfe0ec1c2.

My template looks like this:
```
Schema:
  -
    AttributeDataType: "String"
    Mutable: true
    Name: roles
    StringAttributeConstraints:
      MaxLength: 256
      MinLength: 1
```

And it adds the custom attribute just fine.

**Comment by RossWilliams** [comment]
@yukitaka13-1110 the important part of https://github.com/aws-amplify/amplify-cli/issues/1204#issuecomment-482028480 is that you must modify the cloudformation template before running amplify push. If you have an existing user pool, then you have already run push and it is too late to change Cognito without creating a new user pool.

myCustomAttribute is the name of the attribute. 

Required fields cannot be added after you create a user pool. Optional fields are ok to add inside the Cognito console, but not in Cloudformation template. If you use the Cognito console you can set these optional attributes without creating a new user pool. Additionally, you can write a Custom Resource and use the AWS SDK to add non-required custom properties if you don't want to use the console.

The Cloudformation documentation [here](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-cognito-userpool.html) says that modifying the Schema field requires a replacement of the user pool, meaning it will not update an existing pool.

**Comment by yukitaka13-1110** [comment]
@RossWilliams Oh, I see. I'll try it
Thank you for answering in detail !

**Comment by yukitaka13-1110** [comment]
@RossWilliams I could do it.
Thank you so much !

**Comment by danhumphrey** [comment]
Just FYI for anyone else attempting this, but the example shown in https://github.com/aws-amplify/amplify-cli/issues/1204#issuecomment-482028480 doesn't actually work because custom attributes cannot be `required`

`Reason: Required custom attributes are not supported currently. (Service: AWSCognitoIdentityProviderService; Status Code: 400; Error Code: InvalidParameterException;`

I would happy to be proven wrong if anyone has a way to get `required` custom attributes working?

**Comment by thedgbrt** [comment]
@kaustavghosh06 any progress on this? Would be much cleaner to simply add custom attributes through the amplify cli, and have them be automatically added to various environments.

**Comment by RossWilliams** [comment]
@thedgbrt If you want to automate adding custom attributes today you can do this by creating a Custom Resource and use the aws-sdk to make the right API calls. A side benefit of using the API instead of cloudformation is that updates do not cause your user pool to be replaced.

You would need to use the amplify cli to add a new function, and allow it access to the auth category. Then you will have an environment variable "AUTH_COGNITO${project_name}_USERPOOLID" available in your lambda.

To have your Lambda run when you run amplify push, modify the function's Cloudformation template to add the custom resource inside the "Resources" object:
```json
"CognitoAddCustomAttribute": {
    "Type": "Custom::AddCustomAttribute",
    "Properties": {
        "ServiceToken": {
            "Fn::GetAtt": [
                "LambdaFunction",
	        "Arn"
		]
	},
	"TriggerVersion": 1
    }
},
```

your Lambda would have in its main body (for a numeric attribute):
```javascript
    const aws = require("aws-sdk");
    const identity = new aws.CognitoIdentityServiceProvider({
        apiVersion: "2016-04-18",
        region: process.env.REGION
      });
    await identity
        .addCustomAttributes({
          UserPoolId: process.env.USERPOOL,
          CustomAttributes: [
            {
              Name: "myCustomAttribute",
              AttributeDataType: "Number",
              Mutable: true,
              Required:false ,
              NumberAttributeConstraints: {
                MinValue: "0",
                MaxValue: "1000"
              }
            }
          ]
        })
        .promise()
```
Note that "myCustomAttribute" will be accessible as "custom:myCustomAttribute" for users, Cognito appends the "custom:" part for you.

The template property "TriggerVersion" can be incremented any time you need your function to be re-run in an environment.

I would also modify the "AmplifyResourcesPolicy" to only allow the specific actions you need to add a custom resource.

**Comment by thedgbrt** [comment]
@RossWilliams thank you for taking the time to explain, made me realize amplify was much more powerful than I thought 👍

**Comment by adamelmore** [comment]
@RossWilliams I’m trying to take the custom resource approach you outlined, but my CF stack seems to hang during `amplify push`. The function is run successfully and I see the custom attribute added to my user pool, but the amplify operation never completed (at least not before I cancel ~30 mins in). I think the custom resource may need to leverage `can-response`? I tried with no success. Any tips?

**Comment by RossWilliams** [comment]
@adamelmore, yes, look at 'cfn-response' or 'safe-cfn-custom-resource' to make sure you return a proper response for cloudformation, otherwise your stack with be stuck in an updating state until it times out. Sorry I didn't mention that cloudformation needs a specific response object from your lamba.

**Comment by rfpedrosa** [comment]
I had the same problem as @adamelmore and I solve this by using[ cfn-custom-resource](https://www.npmjs.com/package/cfn-custom-resource) package.

```
exports.handler = async function (event, context) { //eslint-disable-line
const cfnCR = require('cfn-custom-resource');
const { configure, sendSuccess, sendFailure, /*sendResponse,*/ LOG_VERBOSE, /*SUCCESS*/ } = cfnCR;
...

// For Delete requests, immediately send a SUCCESS response.
if (event.RequestType === "Delete") {
      /* Resource successfully created! - async/await */
      const result = await sendSuccess(physicalResourceId, {}, event);
      return result;
}

....

/* Resource successfully created! - async/await */
const result = await sendSuccess(physicalResourceId, {}, event);

or 

await sendFailure(err, event); // if it fails

```

**Comment by SanvirDessai** [comment]
You can create custom attributes during sign up, explained here https://aws.amazon.com/blogs/mobile/aws-amplify-adds-support-for-custom-attributes-in-amazon-cognito-user-pools/

```
import { Auth } from 'aws-amplify'

Auth.signUp({
    'username': 'jdoe',
    'password': 'mysecurepassword#123',
    'attributes': {
        'email': 'me@domain.com',
        'phone_number': '+12135555555',
        'custom:favorite_flavor': 'Cookie Dough'  // custom attribute, not standard
    }
});
```

**Comment by rfpedrosa** [comment]
@SanvirDessai I think this is about the sign-up request use a custom attribute set up in advance on the Cognito pool and **doesn't** automatically create such custom attribute. When I tested, I got a "Attributes did not conform to the schema: custom:X: Attribute does not exist in the schema."

I'm wondering if I miss something. Did you tested/got this working?

**Comment by jkeys-ecg-nmsu** [comment]
@rfpedrosa you are correct, you cannot create a custom attribute that is not part of the Cognito Attributes config. This is a low priority issue for us b/c it's mostly plug-and-play when adding new attributes. 

It would be cool if Amplify could detect the custom attributes provisioned for a project -- which would be trivial if the CLI supported adding custom attributes! -- and plug those into an Amplify-generated migration trigger lambda, so existing attributes are copied over without needing to override the generated migration lambda.

**Comment by SanvirDessai** [comment]
@rfpedrosa I have used this functionality recently. The custom attribute is created and the value is set for the user without creating that custom attribute beforehand. I will try this again and see if I can come up with the steps to replicate this behaviour.

**Comment by JNeim** [comment]
I think it's kind of ridiculous that you can't create a custom Attribute via the CLI after the cognito pool is created and pushed. So basically if you miss something or down the road want to add something in anyway whatsoever you have to completely remake the pool?

**Comment by RossWilliams** [comment]
@JNeim 

> So basically if you miss something or down the road want to add something in anyway whatsoever you have to completely remake the pool? Especially since this has been alive for a year.

No, you can still add custom attributes using the AWS-cli per above comments. You do not need to replace the pool.

**Comment by tomdohnal** [comment]
Isn't creating a `Profile` model in the GraphQL a better approach to add custom fields?

**Comment by Albert-Gao** [comment]
There is another problem here, when using custom attribute, you can only set MaxLength, but on the cognito console, the unit is byte, it is maxByte in the console, and it is 2048, but MaxLength in cloudformation only 256 or something.

**Comment by secretshardul** [comment]
Here's a shell script to [create custom attributes using AWS CLI](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/cognito-idp/add-custom-attributes.html):
```
#!/bin/sh
aws cognito-idp add-custom-attributes \
--user-pool-id <value> \
--custom-attributes \
'[
  {
    "Name": "string",
    "AttributeDataType": "String"|"Number"|"DateTime"|"Boolean",
    "DeveloperOnlyAttribute": true|false,
    "Mutable": true|false,
    "Required": true|false,
    "NumberAttributeConstraints": {
      "MinValue": "string",
      "MaxValue": "string"
    },
    "StringAttributeConstraints": {
      "MinLength": "string",
      "MaxLength": "string"
    }
  }
  ...
]`
```
I couldn't get the custom resource method right, but this worked for me.

**Comment by dylan-westbury** [comment]
> Isn't creating a `Profile` model in the GraphQL a better approach to add custom fields?

By having it in cognito attributes, it's accessible in the claims preventing the additional request that needs to be made to database to access.

**Comment by samuelmburu** [comment]
To anyone who is still researching this in Nov 2021:

I was able to get @RossWilliams solution working using some parts from @rfpedrosa thanks a lot to both of you and this thread.

All was working, but things fell apart after I ran `amplify push -y` after making some changes to my API, the **CognitoAddCustomAttribute** kept running even after trying both `Version` and `TriggerVersion` properties.

The SNS solution seems like the way to go, but I felt like it would add too much time to my project. My solution was to tweak the lambda in the @RossWilliams solution to check if the error was a duplicate - if it was I handled it like a success.

```js
/* Amplify Params - DO NOT EDIT
  AUTH_<user_pool_name>_USERPOOLID
  ENV
  REGION
Amplify Params - DO NOT EDIT */

const aws = require("aws-sdk");
const cfnCR = require('cfn-custom-resource');
const { configure, sendSuccess, sendFailure, /*sendResponse,*/ LOG_VERBOSE, /*SUCCESS*/ } = cfnCR;

/* Increase the logging level */
configure({ logLevel: LOG_VERBOSE });

/**
 * NOTE: this is necessary so user attributes can be updated inside of Cognito without
 * creating a whole new Cognito instance... and nuking old data
 */
exports.handler = async function (event, context, callback) {
  const customAttributes = getCustomAttributes();
  // not sure if this is the right value, but it seems to work ok
  const physicalResourceId = process.env.AUTH_<user_pool_name>_USERPOOLID; 
  const identity = new aws.CognitoIdentityServiceProvider({
    apiVersion: "2016-04-18",
    region: process.env.REGION
  });

  for await (const attribute of customAttributes) {
    try {
      await addCustomAttribute(attribute, identity)

      // if successful do nothing
    } catch (error) {
      // if error break entirely
      // NOTE: Here for visibility - not required by api
      const physicalResourceId = undefined;

      return sendFailure(error, event, callback, context, physicalResourceId);
    }
  }

  return sendSuccess(physicalResourceId, {}, event, callback);
}

/**
 * Adds one custom attribute at a time, making sure that if we attempt to add
 * a duplicate, it will still "show" as successful
 *
 * @param {aws.CognitoIdentityServiceProvider.SchemaAttributeType } attributeDefinition
 * @param {aws.CognitoIdentityServiceProvider} identityProvider
 * @returns
 */
async function addCustomAttribute(attributeDefinition, identityProvider) {
  try {
    await identityProvider
      .addCustomAttributes({
        UserPoolId: process.env.AUTH_<user_pool_name>_USERPOOLID,
        CustomAttributes: [attributeDefinition]
      }).promise();

    return Promise.resolve(null);
  } catch (err) {
    if (err.message?.match(/Existing attribute already/i)) {
      return Promise.resolve(null); // for duplicates return success
    }

    return Promise.reject(err);
  }
}

/**
 * Here to centralize all custom attributes, while ensuring that the most important
 * lambda handler is at the top of this file
 *
 * @returns {aws.CognitoIdentityServiceProvider.SchemaAttributeType[]}
 */
function getCustomAttributes() {
  const customAttributes = [
    {
      Name: "country",
      AttributeDataType: "String",
      Mutable: true,
      // Required: true, // NOTE: commenting out because required custom attributes are not supported
      StringAttributeConstraints: {
        MinLength: "1",
        MaxLength: "140"
      }
    }
	// more attributes?
  ];

  return customAttributes;
}
```

Hopefully, this helps someone else.

--
I did walk into 2 "gotchas" with the **CognitoAddCustomAttribute**:†

1. when defining it, I had to hard code the lambda Arn
```json
"CognitoAddCustomAttribute": {
    "Type": "Custom::AddCustomAttribute",
    "ServiceToken": {
          "Fn::If": [
            "HasEnvironmentParameter",
            {
              "Fn::Sub": [
                "arn:aws:lambda:${AWS::Region}:${AWS::AccountId}:function:<lambda_name>-${env}",
                {
                  "env": {
                    "Ref": "env"
                  }
                }
              ]
            },
            {
               "Fn::Sub": [
                "arn:aws:lambda:${AWS::Region}:${AWS::AccountId}:function:<lambda_name>",
                {}
              ]
            }
          ]
        },
	"TriggerVersion": 1
    }
}
```
2. When deploying, I had to first deploy the lambda, and then the updates to the API category

These issues were more than likely due to me not knowing how to most effectively use cloudformation and how the parameters/vars work -- but it works and is resilient enough for my purposes.

**Comment by KalvinWei** [comment]
> @yukitaka13-1110 An example [83476b9?short_path=ac1bfc4#diff-ac1bfc471563fa82f8882dccfe0ec1c2](https://github.com/aws-amplify/amplify-cli/commit/83476b944852ffa000875074e200b4fd5ec6f88c?short_path=ac1bfc4#diff-ac1bfc471563fa82f8882dccfe0ec1c2).
> 
> My template looks like this:
> 
> ```
> Schema:
>   -
>     AttributeDataType: "String"
>     Mutable: true
>     Name: roles
>     StringAttributeConstraints:
>       MaxLength: 256
>       MinLength: 1
> ```
> 
> And it adds the custom attribute just fine.

Hi, @jarrettj . Do we need to modify the *parameters.json* file as well?

**Comment by undefobj** [comment]
@josefaidt until the feature is implemented lets add a docs example of using overrides for this capability
